RFC: [gssproxy] Improved systemd integration

Alphix opened the pull request:
These patches improve the systemd integration and also hardens the systemd `.service` unit.

Lightly tested, so consider an RFC for now. I want some feedback if this is desirable before I spend more time on testing, etc.

Alphix commented:
@simo5 not sure if you get notified about draft PRs, if not, ping :)

simo5 commented:
I get notified, but feel free to ping if I miss them, I sometimes get busy elsewhere and then forget to check back.

simo5 commented on src/gp_init.c:
```diff
@@ -255,15 +256,21 @@ int init_userproxy_socket(struct gssproxy_ctx *gpctx)
+    struct gssproxy_ctx *gpctx = verto_get_private(ev);
     struct gp_config *new_config, *old_config;
 
-    gpctx = verto_get_private(ev);
```
Leave this as it was please.

simo5 commented on src/gp_init.c:
```diff
@@ -255,15 +256,21 @@ int init_userproxy_socket(struct gssproxy_ctx *gpctx)
     if (!new_config) {
+        sd_notifyf(0, "READY=1\n"
+                   "STATUS=Running, %i service(s) configured\n",
+                   gpctx->config->num_svcs);
```
Shouldn't we notify of an error status/warning of some kind, or does sd_notify not deal with thst ?

simo5 replied:
Maybe just add "(Failed to re-read config)" in the STATUS part?

Alphix replied:
Good idea, will fix

Alphix replied:
> Shouldn't we notify of an error status/warning of some kind, or does sd_notify not deal with thst ?

I don't think it's necessary to check the return value of `sd_notify` because there's not much that can be done about it....if `sd_notify` doesn't work, the system is either truly borked (and `sd_notifyf` is the least of your problems) or there's code somewhere that is unsetting the relevant environment variable. The man page for `sd_notifyf` says "`it is generally recommended to ignore the return value of this call`"

simo5 commented on src/gp_util.c:
```diff
@@ -7,9 +7,38 @@
+{
+    struct timespec ts;
+
+    if (clock_gettime(CLOCK_MONOTONIC, &ts) < 0) {
```
does this really need to be the monotonic clock ?
Seem it is used only for logging, so perhaps the actual system time is a better identifier here?

Alphix replied:
Yes, it's the protocol for `Type=notify-reload`, documented in `man systemd.service(5)` (on a sufficiently recent system). Basically I wrote it like this so that the systemd integration is backwards and forwards compatible (right now the patches only use `Type=notify` in the `.service` file for backwards compatibility, but the resulting code itself supports the newer `Type=notify-reload` protocol out-of-the-box)

simo5 commented on systemd/gssproxy.service.in:
```diff
@@ -5,30 +5,50 @@ After=syslog.target network.target
+PrivateIPC=yes
+# Blocks access to /home which may hold ccaches
+PrivateUsers=no
+ProtectHome=read-only
```
This is wrong, caches may need to be writable.

Alphix replied:
Ok, I wasn't sure about that one, I'll fix that

Alphix replied:
`ProtectSystem=strict` already makes the whole FS read-only (except `/var/lib/gsssproxy` which is made RW via the `StateDirectory=` stanza)...so, we'd need to create `BindPaths=` exceptions for any location where writeable ccaches may be expected....when/where/under which circumstances are there KRB5 ccaches under `/home` and/or `/root`? All systems I've used have ccaches under `/tmp`...(unless the kernel keyring or something like `sssd`s KCM are used)...I need some pointers

simo5 replied:
System specific, the traditional /tmp is considerd broken due TokTou, so on some systems people use their home directories. But there is also the simple case of a user creating a ccache anywhere by just setting KRB5CCNAME in the process (and adding a corresponding gssproxy configuration stanza to point to it).
The DIR type at some point pointed to /run/user directories.
The Keyring or KCM cases are the easy one to handle (guess why I add both those facilities in krb5 and SSSD :-) 

I do not think I can change this now, it would break any user that decided to place writable cacches in a non standard place.

Alphix replied:
Gotcha, so your concern is systems with customised ccache paths via e.g. `KRB5CCNAME` or a customised `/etc/krb5.conf` along these lines:

```
[libdefaults]
default_cc_name = FILE:${homedir}/.krb5cc
```

So which paths would you like to see as being made read-writeable by default (note, it can still be changed locally via systemd drop-ins)? The whole `/root`, `/home` and `/run/user` hierarchy? Or could we default to a more restricted subset and document that systems with such changes should also create a suitable drop-in...?

simo5 replied:
I think the regular paths a user can write in should be allowed, so /home /root /run/user sounds reasonable.
And yes the rest via drop-in.

simo5 commented on systemd/gssproxy.service.in:
```diff
@@ -5,30 +5,50 @@ After=syslog.target network.target
+ProtectKernelLogs=yes
+ProtectControlGroups=yes
+RestrictAddressFamilies=AF_UNIX AF_LOCAL
+RestrictNamespaces=yes
```
Not sure this is correct, when there are multiple name spaces created by containers we want the gssproxy to be working within the correct namespace too if they mount the socket.
I haven't fully read what RestrictNamespaces do, so I may just be grasping at straws here.

Alphix replied:
I'm fairly certain that the patch will break some systems given all the hardening options. I'll have a look at the containers/namespaces scenario

simo5 replied:
Running nfs server correctly for different namespaces is one of the goals of gssproxy.
If this needs service file created for specific namespaces we can document it, and leave the main service file more restricted as well.
But if we can make the main service file compatible with multiple nfs servers/clients in different namespaces we should try.

Alphix replied:
> Running nfs server correctly for different namespaces is one of the goals of gssproxy. If this needs service file created for specific namespaces we can document it, and leave the main service file more restricted as well. But if we can make the main service file compatible with multiple nfs servers/clients in different namespaces we should try.

Does `gssproxy` muck around with namespaces or it is just launched *in* a given namespace? After some quick grepping I can't find code that would actually change namespaces (`unshare`, `setns`, `clone`)....? If so, the `RestrictNamespaces=` should be ok...since it's about restricting `ns` *changes*, not `ns` support per se...

simo5 reviewed (commented):
As a general concept I like this PR.
But need to look at some details more closely.

Alphix commented:
> As a general concept I like this PR.

Good, do you want to keep the (incomplete, because there's no `/etc/init.d/gssproxy` file) support for non-systemd systems? Because if the answer is no, there's a whole bunch of code that can be simplified/removed....

simo5 commented:
> > I think the first patch will break older systems. The systemd detection is complicated because older versions had a different package split, but I forgot the deatils and may well be wrong.
> 
> I think you might be referring to systemd 209? That's covered in the patch message...that was when the `libsystemd-*.so` libraries were merged into `libsystemd.so`...and that was in 2014.
> 
> Or are you referring to something else?

I was, but if RHEL-8 is not affected I wthink we can definitely change it.

simo5 commented:
> > As a general concept I like this PR.
> 
> Good, do you want to keep the (incomplete, because there's no `/etc/init.d/gssproxy` file) support for non-systemd systems? Because if the answer is no, there's a whole bunch of code that can be simplified/removed....

I want to still make it possible to have this work w/o systemd, but I do not care for shipping init files myself.

Alphix commented:
>> I think you might be referring to systemd 209?

> I was, but if RHEL-8 is not affected I wthink we can definitely change it.

RHEL8 was (from what I've understood from a quick Google search) released in 2019 and uses systemd v239? If so, shouldn't be a problem...

Alphix commented:
> > ...do you want to keep the...support for non-systemd systems?...

> I want to still make it possible to have this work w/o systemd, but I do not care for shipping init files myself.

Ok, good, then the current state should be fine...

Alphix commented:
@simo5 : Ok, rebased version pushed, should address all feedback you've provided so far...by default, `/tmp`, `/root`, `/home`, `/run/user` and `/var/gssproxy` should now be read-writable to `gssproxy`. I've added a comment on how to further restrict this using a drop-in. Hopefully that can be reversed in the future (i.e. to make `/tmp`, `/root`, `/home` and `/run/user` RW access opt-in instead of opt-out).

simo5 commented:
Thanks looks good, therefore I marked it ready and merging
